enforcer.go

```diff
@@ -419,78 +421,50 @@ func (e *Enforcer) enforce(matcher string, explains *[]string, rvals ...interfac
 				len(rvals),
 				rvals)
 		}
-		for i, pvals := range e.model["p"]["p"].Policy {
-			// log.LogPrint("Policy Rule: ", pvals)
-			if len(e.model["p"]["p"].Tokens) != len(pvals) {
-				return false, fmt.Errorf(
-					"invalid policy size: expected %d, got %d, pvals: %v",
-					len(e.model["p"]["p"].Tokens),
-					len(pvals),
-					pvals)
-			}
-
-			parameters.pVals = pvals
-
-			if hasEval {
-				ruleNames := util.GetEvalValue(expString)
-				var expWithRule = expString
-				for _, ruleName := range ruleNames {
-					if j, ok := parameters.pTokens[ruleName]; ok {
-						rule := util.EscapeAssertion(pvals[j])
-						expWithRule = util.ReplaceEval(expWithRule, rule)
-					} else {
-						return false, errors.New("please make sure rule exists in policy when using eval() in matcher")
-					}
 
-					expression, err = govaluate.NewEvaluableExpressionWithFunctions(expWithRule, functions)
+		if e.parallel {
+			errs := make(chan error)
+			var wg sync.WaitGroup
+
+			for i, pvals := range e.model["p"]["p"].Policy {
+				wg.Add(1)
+				go func(i int, pvals []string, parameters enforceParameters, expression *govaluate.EvaluableExpression) {
+					defer wg.Done()
+					parameters.pVals = pvals
+					var err error
+					policyEffects[i], matcherResults[i], err = e.evaluatePolicy(parameters, expression, expString, functions)
 					if err != nil {
-						return false, fmt.Errorf("p.sub_rule should satisfy the syntax of matcher: %s", err)
+						errs <- err
 					}
-				}
-
+				}(i, pvals, parameters, expression)
 			}
 
-			result, err := expression.Eval(parameters)
-			// log.LogPrint("Result: ", result)
+			go func() {
+				wg.Wait()
+				close(errs)
+			}()
+
+			var err error
```

Comment:
I don't thinks this is correct.  We need wait evaluatePolicy is done then check errors.

Reply:
please use `ctx, cancel := context.WithCancel(context.Background())` to control logic.

Reply:
I update it according to your example. But, I'm not really sure if I understand your review correctly. Kindly check it again :)